### PR TITLE
Dynamic Certificate Serving - Controller

### PR DESCRIPTION
**Pull request by hnrytrn**
What this PR does / why we need it:

This PR is a part of adding dynamic certificate serving functionality (Shopify/edgescale#515). When dynamic configuration mode is on, the controller will JSON encode the certificates and post it to an internal endpoint.

@Shopify/edgescale

### CODE REVIEW

**Comment by andrewlouis93** [comment]
@hnrytrn At a once over, this looks pretty decent - will do a more thorough review tomorrow though. Can you PR this upstream instead? Let's merge there and then re-sync to our fork.

File: internal/ingress/controller/controller.go [hunk 1]
```diff
@@ -197,6 +197,14 @@ func (n *NGINXController) syncIngress(interface{}) error {
 			} else {
 				glog.Warningf("Dynamic reconfiguration failed: %v", err)
 			}
+
+			// Configuring certificates
```
Comment (fmejia97) [inline internal/ingress/controller/controller.go]: @hnrytrn I would suggest moving the `configureCerts` call  and error checking to be inside `configureDynamically`.  `configureDynamically` holds the sync logic when dynamic configuration is enabled. Currently, `configureDynamically` only posts `Backend` data to the lua `configuration` module to keep it up to date with the latest `Backend` changes. With dynamic certs, we are now introducing another resource to post to `configuration` lua module and keep track of.  Therefore  `configureDynamically` would encapsulate the logic behind keeping `lua` configuration up to speed. It would be a good idea to move the current contents of `configureDynamically` to a function called `configureBackends` and then you could just call `configureBackends()` and `configureCerts()` inside. What are your thoughts?

Comment (hnrytrn) [reply internal/ingress/controller/controller.go]: Yeah, this is a good idea, I'll update the PR.

File: internal/ingress/controller/controller.go [hunk 2]
```diff
@@ -1031,6 +1031,12 @@ func (n *NGINXController) createServers(data []*extensions.Ingress,
 
 			servers[host].SSLCert = *cert
 
+			if n.cfg.DynamicConfigurationEnabled {
+				// Certificates won't be written on disk during dynamic configuration mode
+				servers[host].SSLCert.PemFileName = defaultPemFileName
+				servers[host].SSLCert.PemSHA = defaultPemSHA
+			}
```
Comment (andrewlouis93) [inline internal/ingress/controller/controller.go]: Could we add some integration tests for this new behaviour? 

A good reference starting point is looking at this file: https://github.com/kubernetes/ingress-nginx/blob/ee46f486c77665096f366eb215aa135032291b5a/test/e2e/lua/dynamic_configuration.go

File: internal/ingress/controller/store/backend_ssl.go [hunk 3]
```diff
@@ -148,7 +154,7 @@ func (s k8sStore) checkSSLChainIssues() {
 			continue
 		}
 
-		data, err := ssl.FullChainCert(secret.PemFileName, s.filesystem)
+		data, err := ssl.FullChainCert([]byte(secret.PemCertKey))
```
Comment (andrewlouis93) [inline internal/ingress/controller/store/backend_ssl.go]: @hnrytrn 👋 Could you explain why this change was necessary? I think I'm missing a beat.

File: internal/net/ssl/ssl.go [hunk 4]
```diff
@@ -178,6 +185,107 @@ func AddOrUpdateCertAndKey(name string, cert, key, ca []byte,
...
 		CN:          cn.List(),
 		ExpireTime:  pemCert.NotAfter,
+		PemCertKey:  string(pemCerts),
+	}
+
+	return s, nil
+}
+
+// CreateSSLCert creates an SSLCert and avoids writing on disk unless there is a CA
+func CreateSSLCert(name string, cert, key, ca []byte, fs file.Filesystem) (*ingress.SSLCert, error) {
```
Comment (andrewlouis93) [inline internal/net/ssl/ssl.go]: Unit test missing; see: https://github.com/kubernetes/ingress-nginx/blob/master/internal/net/ssl/ssl_test.go

File: internal/ingress/controller/store/backend_ssl.go (hunk 3 above)
Comment (hnrytrn) [reply internal/ingress/controller/store/backend_ssl.go]: One of the changes I added is that the PEM certificate isn't written on disk during dynamic configuration mode, so `PemFileName` would be empty.

File: internal/ingress/controller/nginx.go [hunk 5]
```diff
@@ -784,6 +816,48 @@ func configureDynamically(pcfg *ingress.Configuration, port int) error {
...
+		servers = append(servers, &ingress.Server{
+			Hostname: server.Hostname,
+			SSLCert: ingress.SSLCert{
+				PemCertKey: server.SSLCert.PemCertKey,
+			},
+		})
+	}
+
+	buf, err := json.Marshal(servers)
+
```
Comment (andrewlouis93) [inline internal/ingress/controller/nginx.go]: nitpick, let's get rid of the new lines between err assignment and returning. Also have you run `gofmt`? 
```
_, err = ...
if err != nil {
   return err
}
```

File: internal/ingress/controller/store/backend_ssl.go (hunk 3 above)
Comment (andrewlouis93) [reply internal/ingress/controller/store/backend_ssl.go]: Ah of course :)

File: internal/net/ssl/ssl.go [hunk 6]
```diff
@@ -178,6 +185,107 @@ func AddOrUpdateCertAndKey(name string, cert, key, ca []byte,
 		PemSHA:      file.SHA1(pemFileName),
 		CN:          cn.List(),
 		ExpireTime:  pemCert.NotAfter,
+		PemCertKey:  string(pemCerts),
+	}
+
+	return s, nil
+}
+
+// CreateSSLCert creates an SSLCert and avoids writing on disk unless there is a CA
```
Comment (ElvinEfendi) [inline internal/net/ssl/ssl.go]: > avoids writing on disk unless there is a CA

why is this the case?

Comment (hnrytrn) [reply internal/net/ssl/ssl.go]: When there is a CA, `CAFileName` gets created and in `nginx.tmpl` `ssl_client_certificate` is set to this. It doesn't look like OpenResty can set `ssl_client_certificate` dynamically so we have to write this on disk.

File: internal/net/ssl/ssl.go [hunk 7]
```diff
@@ -162,13 +163,19 @@ func AddOrUpdateCertAndKey(name string, cert, key, ca []byte,
 		caFile.Write([]byte("\n"))
 		defer caFile.Close()
 
+		pemCertKey, err := fs.ReadFile(pemFileName)
```
Comment (ElvinEfendi) [inline internal/net/ssl/ssl.go]: at https://github.com/Shopify/ingress/pull/68/files#diff-a3867fb2b922c407f56282c077f2177dR102 you avoid calling this function when dynamic cert mode enabled, why do you need this change here? Can`PemCertKey` not be empty when dynamic cert mode is not on?

File: internal/ingress/controller/nginx.go [hunk 8]
```diff
@@ -784,6 +803,46 @@ func configureDynamically(pcfg *ingress.Configuration, port int) error {
 	return nil
 }
 
+// configureCertificates JSON encodes certificates and POSTs it to an internal HTTP endpoint
+// that is handled by Lua
+func configureCertificates(pcfg *ingress.Configuration, port int) error {
```
Comment (ElvinEfendi) [inline internal/ingress/controller/nginx.go]: let's call this function inside `configureDynamically` and also extract the POST logic into a helper function

**Comment by hnrytrn** [comment]
Move PR upstream - https://github.com/kubernetes/ingress-nginx/pull/2923
